aten/src/ATen/native/mobile/cpu/internal/Add.cpp

```diff
@@ -0,0 +1,244 @@
+#include <ATen/native/mobile/cpu/Engine.h>
+
+#ifdef USE_XNNPACK
+
+#include <ATen/native/mobile/cpu/internal/Allocator.h>
+#include <ATen/native/mobile/cpu/internal/Common.h>
+
+namespace at {
+namespace native {
+namespace mobile {
+namespace cpu {
+namespace internal {
+namespace add {
+
+struct Context final {
+  Operator add_op;
+
+  static constexpr float kMin = -std::numeric_limits<float>::infinity();
+  static constexpr float kMax = std::numeric_limits<float>::infinity();
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32 adds.
+
+bool available(
```

Comment:
Seems this is kinda new dispatching mechanism in ATen. AFAIN the existing ones are:

1. New backend type, e.g.: CPU, CUDA, QuantizedCPU, MkldnnCPU, etc. It works best for cases where dispatching criteria are based on some intrinsic property of input tensor, e.g. TensorId (CPU Tensor v.s. CUDA Tensor have different tensor id and different storage type). This type of dispatching is managed by aten codegen and several yaml files containing metadata;

2. New engine type - the only case is probably FBGEMM v.s. QNNPACK which both work for QuantizedCPU backend. Both share pretty much the same APIs, pack/unpack workflow, can take the same type of input tensor, and can run on the same CPU architecture, thus we introduce this new concept of engine to do dispatching dynamically;

3. instruction set specific function dispatch (aten/src/ATen/native/DispatchStub.h) - a compile time switch for ops in native/cpu/* based on cpu capability (avx2, avx, etc);

4. Ad-hoc logic - a classic example is Convolution.cpp - based on a set of compile time flags (e.g. AT_NNPACK_ENABLED) and runtime properties (e.g. is_depthwise / check_cudnn_depthwise_workload) it calls different conv kernels;

I think this new available() / usable() APIs are strictly better than approach 4) (what we took for nnpack / depthwise3x3 kernel integration), thanks for making the effort to clean it up! Looks like we are going to call these APIs in places where 4) happens today to dispatch to XNNPACK kernels when proper?

It seems fine to be but I'll let @ezyang @gchanan @dzhulgakov also take a look. Do we have a long term plan / guideline for new kernel integration work (like this one and upcoming mobile GPU)?

Reply:
> I'd think about these two questions:
> 
> Seems currently mobile/*.cpp are mostly for XNNPACK integration. Do we have plan to add more mobile specific integration in future?
> Do we want to use XNNPACK in non-mobile cases in the future?
> 
> The current setup works best for Yes-No case. If the answers are No-Yes, then maybe we should consider neutralizing the naming to be less mobile specific?
> 

@ljk53 I was planning on adding the mobile gpu backend as a separate folder right next to cpu to consolidate all mobile backends together.  I agree that neither XNNPACK, nor the GPU backend under work, is mobile specific.  I am very open to changing the folder structure to how you guys think is best.  Suggestions are welcome.

Reply:
Yeah what I was trying to get at is to keep the selection logic close to the actual operator.  If we decide to expose these operators through _convolution and other native implementations, which is not ideal from a performance perspective but still better than NNPACK in at least some cases, then those native functions would call available() and usable().  The difference being that available() gates op creation, checking all the data that is available at the time of creation, while usable checks everything else at execution time, mainly parameters that are not available until the very time we are calling forward - such as the input tensor.

If exposed through Python modules, like I did in the previous PR, these functions end up being called from the Python module constructor, or forward() respectively, and throw an exception if some condition is not met.  Marshaling C++ exceptions to Python requires some extra work it seems though.

Reply:
OK so to comment on the current state of this issue - following yesterday's meeting we decided we will expose this functionality through native_functions.  available() and run() will be used internally in this file and the like, for asserts and error checking.